// use-script/dist/cjs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeFailedScript = exports.getScriptTag = exports.loadScript = exports.setTestLoadMode = void 0;
var tslib_1 = require("tslib");
var testLoadFn;
exports.setTestLoadMode = function (load) {
    if (process.env.NODE_ENV !== 'production') {
        testLoadFn = load;
    }
};
function loadScript(_a) {
    var src = _a.src, attr = tslib_1.__rest(_a, ["src"]);
    if (process.env.NODE_ENV !== 'production' && testLoadFn) {
        var script_1 = document.createElement('script');
        script_1.src = src;
        Object.keys(attr).forEach(function (key) { return (script_1[key] = attr[key]); });
        document.body.appendChild(script_1);
        return testLoadFn()
            .then(function () { return ({
            success: true,
            element: script_1,
            script: src,
            error: null,
        }); })
            .catch(function () { return ({
            success: false,
            element: script_1,
            script: src,
            error: new ErrorEvent('test'),
        }); });
    }
    return new Promise(function (resolve) {
        var script = document.createElement('script');
        script.src = src;
        Object.keys(attr).forEach(function (key) { return (script[key] = attr[key]); });
        script.addEventListener('load', function () {
            return resolve({
                element: script,
                script: src,
                success: true,
                error: null,
            });
        });
        script.addEventListener('error', function (e) {
            return resolve({
                element: script,
                script: src,
                success: false,
                error: e,
            });
        });
        document.body.appendChild(script);
    });
}
exports.loadScript = loadScript;
function getScriptTag(script) {
    if (typeof document === 'undefined') {
        return;
    }
    return document.querySelector("script[src='" + script + "']");
}
exports.getScriptTag = getScriptTag;
function removeFailedScript(script) {
    var node = document.querySelector("script[src='" + script + "']");
    if (node && node.parentNode) {
        node.parentNode.removeChild(node);
    }
}
exports.removeFailedScript = removeFailedScript;
